File: ui/packages/ui/src/search/engine/chars.ts
/** Unicode 标量扫描。不进模块公开面。 */

export function loweredChars(text: string): string[] {
  return [...text.toLowerCase()];
}

export function findChars(hay: readonly string[], needle: readonly string[], from: number): number | null {
  if (needle.length === 0) return Math.min(from, hay.length);
  if (needle.length > hay.length) return null;
  const last = hay.length - needle.length;
  for (let i = from; i <= last; i += 1) {
    let ok = true;
    for (let j = 0; j < needle.length; j += 1) {
      if (hay[i + j] !== needle[j]) {
        ok = false;
        break;
      }
    }
    if (ok) return i;
  }
  return null;
}

File: core/yohu-search/src/chars.rs
//! Unicode 标量扫描。不进公开面。

pub fn lowered_chars(text: &str) -> Vec<char> {
    text.to_lowercase().chars().collect()
}

pub fn find_chars(hay: &[char], needle: &[char], from: usize) -> Option<usize> {
    if needle.is_empty() {
        return Some(from.min(hay.len()));
    }
    if needle.len() > hay.len() {
        return None;
    }
    let last = hay.len() - needle.len();
    let mut i = from;
    while i <= last {
        if hay[i..i + needle.len()] == needle[..] {
            return Some(i);
        }
        i += 1;
    }
    None
}
